fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,428 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+		goto out;
+	}
+
+	/* This test is only setup to support selected providers. */
+	if (!strstr(hints->fabric_attr->prov_name, "ofi_rxm")) {
+		FT_DEBUG("%s not support for this test\n",
+			 hints->fabric_attr->prov_name);
+		return FI_SUCCESS;
+	}
+
+	hints->domain_attr->name = strdup((*info)->domain_attr->name);
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	if (!hints->domain_attr->name) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	/* Pass in the majority of the require test capabilities as optional.
+	 * This capabilites should be set in the return fi_info::caps. In
+	 * addition, since user hint domain_attr::caps, tx_attr::caps, and
+	 * rx_attr::caps are zero, these returns caps should be set based on
+	 * fi_info::caps
+	 */
+	hints->caps = FI_TAGGED;
+	hints->optional_caps = FI_ATOMIC;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, FI_ATOMIC, FI_ATOMIC,
+					   FI_ATOMIC);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	/* For RxM, FI_ATOMIC used as optional_cap fi_getinfo() input should not
+	 * be returned in caps output since performance could be impacted.
+	 */
+	ret = test_invalid_cap_info(*info, FI_ATOMIC, FI_ATOMIC, FI_ATOMIC);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Verify that using domain_attr::optional_caps, tx_attr::optional_caps,
+	 * and rx_attr::optional_caps works as expected.
+	 */
+	hints->caps = required_caps;
+	hints->optional_caps = 0;
+	hints->tx_attr->caps = FI_TAGGED;
+	hints->rx_attr->caps = FI_TAGGED;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, 0, 0);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	hints->tx_attr->optional_caps = FI_ATOMIC;
+	hints->rx_attr->optional_caps = FI_ATOMIC;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, FI_ATOMIC, FI_ATOMIC);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	ret = test_invalid_cap_info(*info, 0, FI_ATOMIC, FI_ATOMIC);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+out:
+	return ret;
+}
+
+static int test_ignore_optional_cap(char *node, char *service, uint64_t flags,
+				    struct fi_info *hints,
+				    struct fi_info **info)
+{
+	uint64_t required_caps = FI_MSG | FI_TAGGED;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 19), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
```

Comment:
I am going to update these tests to use more of the fabtest test status and error reporting infrastructure